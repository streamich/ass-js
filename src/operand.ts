import {UInt64} from './util';
import {Expression} from './expression';
import Label from './Label';


export enum SIZE {
    ANY     = -1,       // Any size.
    NONE    = 0,        // Either unknown or not specified yet.
    B       = 8,        // byte
    W       = 16,       // word
    D       = 32,       // double word = 2x word
    Q       = 64,       // quad word = 4x word
    T       = 80,       // ten = 10 bytes
    O       = 128,      // octa word = 8x word
    H       = 256,      // hexadeca word = 16x word
    I       = 512,      // dItriaconta word = 32x word
    X       = SIZE.O,   // xmm register
    Y       = SIZE.H,   // ymm register
    Z       = SIZE.Y,   // zmm register
    A       = 1024,     // amm register
}

export enum SIZEB {
    B1      = SIZE.B,
    B2      = SIZE.W,
    B4      = SIZE.D,
    B8      = SIZE.Q,
    B16     = SIZE.O,
    B32     = SIZE.H,
    B64     = SIZE.I,
}


// 64-bit numbers
export type number64 = [number, number];
export function isNumber64(num) {
    if((num instanceof Array) && (num.length === 2) && (typeof num[0] === 'number') && (typeof num[1] === 'number')) return true;
    else return false;
}

// 128-bit numbers
export type number128 = [number, number, number, number];
export function isNumber128(num) {
    if((num instanceof Array) && (num.length === 4) &&
        (typeof num[0] === 'number') && (typeof num[1] === 'number') &&
        (typeof num[1] === 'number') && (typeof num[2] === 'number')) return true;
    else return false;
}

function isNumberOfDoubles(doubles, num) {
    if(!(num instanceof Array) || (num.length !== doubles)) return false;
    for(var j = 0; j < doubles; j++)
        if(typeof num[j] !== 'number') return false;
    return true;
}

// 256-bit numbers
// export type number256 = [number, number, number, number, number, number, number, number];
// export function isNumber256(num) { return isNumberOfDoubles(8, num); }

// 512-bit numbers
// export type number512 = [number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number];
// export function isNumber512(num) { return isNumberOfDoubles(16, num); }

// AVX-512 extension says registers will be "at least" 512 bits, so can be 1024 bits and maybe even 2048 bits.
// export type number1024 = [number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number];
// export type number2048 = [number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number, number];
// export function isNumber1024(num) { return isNumberOfDoubles(32, num); }
// export function isNumber2048(num) { return isNumberOfDoubles(64, num); }


// Combined type of all basic "numbers".
export type Tnumber = number|number64|number128;
// export type Tnumber = number|number64|number128|number256|number512|number1024|number2048;
export function isTnumber(num) {
    if(typeof num === 'number') return true;
    else if(isNumber64(num))    return true;
    // else if(isNumber128(num))   return true;
    // else if(isNumber256(num))   return true;
    // else if(isNumber512(num))   return true;
    // else if(isNumber1024(num))  return true;
    // else                        return isNumber2048(num);
    else                        return isNumber128(num);
}


export type TUiOperand           = Register|Memory|Tnumber|Relative|Expression;
export type TUiOperandNormalized = Register|Memory|Tnumber|Relative; // `Expression` gets converted to `Relative`.


// General operand used in our assembly "language".
export abstract class Operand {
    // Size in bits.
    size: SIZE = SIZE.ANY;

    // Convenience method to get `Register` associated with `Register` or `Memory`.
    reg(): Register {
        return null;
    }

    isRegister(): boolean {
        return this instanceof Register;
    }

    isMemory(): boolean {
        return this instanceof Memory;
    }

    toNumber(): Tnumber {
        return 0;
    }

    toString(): string {
        return '[operand]';
    }
}


// ## Constant
//
// Constants are everything where we directly type in a `number` value.
export class Constant extends Operand {

    static sizeClass(value) {
        if((value <= 0x7f) && (value >= -0x80))             return SIZE.B;
        if((value <= 0x7fff) && (value >= -0x8000))         return SIZE.W;
        if((value <= 0x7fffffff) && (value >= -0x80000000)) return SIZE.D;
        return SIZE.Q;
    }

    static sizeClassUnsigned(value) {
        if(value <= 0xff)           return SIZE.B;
        if(value <= 0xffff)         return SIZE.W;
        if(value <= 0xffffffff)     return SIZE.D;
        return SIZE.Q;
    }

    value: number|number64 = 0;

    // Each byte as a `number` in reverse order.
    octets: number[] = [];

    signed: boolean = true;

    constructor(value: Tnumber = 0, signed = true) {
        super();
        this.signed = signed;
        this.setValue(value);
    }

    setValue(value: number|number64) {
        if(isNumber64(value)) {
            this.setValue64(value as number64);
        } else if(typeof value === 'number') {
            var clazz = this.signed ? Constant.sizeClass(value) : Constant.sizeClassUnsigned(value);
            /* JS integers are 53-bit, so split here `number`s over 32 bits into [number, number]. */
            if(clazz === SIZE.Q) this.setValue64([UInt64.lo(value), UInt64.hi(value)]);
            else                 this.setValue32(value);
        } else
            throw TypeError('Constant value must be of type Tnumber.');
    }

    protected setValue32(value: number) {
        var size = this.signed ? Constant.sizeClass(value) : Constant.sizeClassUnsigned(value);
        this.size = size;
        this.value = value;
        this.octets = [value & 0xFF];
        if(size > SIZE.B) this.octets[1] = (value >> 8) & 0xFF;
        if(size > SIZE.W) {
            this.octets[2] = (value >> 16) & 0xFF;
            this.octets[3] = (value >> 24) & 0xFF;
        }
    }

    protected setValue64(value: number64) {
        this.size = SIZE.Q;
        this.value = value;
        var [lo, hi] = value;
        this.octets = [
            (lo) & 0xFF,
            (lo >> 8) & 0xFF,
            (lo >> 16) & 0xFF,
            (lo >> 24) & 0xFF,
            (hi) & 0xFF,
            (hi >> 8) & 0xFF,
            (hi >> 16) & 0xFF,
            (hi >> 24) & 0xFF,
        ];
    }

    protected setValue128(value: number128) {
        this.size = SIZE.O;
        this.value = value;
        var [b0, b1, b2, b3] = value;
        this.octets = [
            (b0) & 0xFF,
            (b0 >> 8) & 0xFF,
            (b0 >> 16) & 0xFF,
            (b0 >> 24) & 0xFF,
            (b1) & 0xFF,
            (b1 >> 8) & 0xFF,
            (b1 >> 16) & 0xFF,
            (b1 >> 24) & 0xFF,
            (b2) & 0xFF,
            (b2 >> 8) & 0xFF,
            (b2 >> 16) & 0xFF,
            (b2 >> 24) & 0xFF,
            (b3) & 0xFF,
            (b3 >> 8) & 0xFF,
            (b3 >> 16) & 0xFF,
            (b3 >> 24) & 0xFF,
        ];
    }

    zeroExtend(size) {
        if(this.size === size) return;
        if(this.size > size) throw Error(`Already larger than ${size} bits, cannot zero-extend.`);
        var missing_bytes = (size - this.size) / 8;
        this.size = size;
        for(var i = 0; i < missing_bytes; i++) this.octets.push(0);
    }

    signExtend(size) {
        if(this.size === size) return;
        if(this.size > size) throw Error(`Already larger than ${size} bits, cannot zero-extend.`);

        // TODO: Make it work with 128-bit numbers too, below.

        // We know it is not number64, because we don't deal with number larger than 64-bit,
        // and if it was 64-bit already there would be nothing to extend.
        var value = this.value as number;

        if(size === SIZE.Q) {
            this.setValue64([UInt64.lo(value), UInt64.hi(value)]);
            return;
        }

        this.size = size;
        this.octets = [value & 0xFF];
        if(size > SIZE.B) this.octets[1] = (value >> 8) & 0xFF;
        if(size > SIZE.W) {
            this.octets[2] = (value >> 16) & 0xFF;
            this.octets[3] = (value >> 24) & 0xFF;
        }
    }

    extend(size: SIZE) {
        if(this.signed) this.signExtend(size);
        else this.zeroExtend(size);
    }

    fitsSize(num: Tnumber) {
        var size = this.signed ? Constant.sizeClass(num) : Constant.sizeClassUnsigned(num);
        return size <= this.size;
    }

    toString() {
        var str = '';
        for(var i = this.octets.length - 1; i >= 0; i--) {
            var oct = this.octets[i];
            str += oct > 0xF ? oct.toString(16).toUpperCase() : '0' + oct.toString(16).toUpperCase();
        }
        return '0x' + str;
    }
}


export class Immediate extends Constant {
    static factory(size, value: number|number64 = 0, signed = true) {
        switch(size) {
            case SIZE.B:    return new Immediate8(value, signed);
            case SIZE.W:    return new Immediate16(value, signed);
            case SIZE.D:    return new Immediate32(value, signed);
            case SIZE.Q:    return new Immediate64(value, signed);
            default:        return new Immediate(value, signed);
        }
    }

    static isImmediateClass(Clazz: any) {
        return Clazz.name.indexOf('Immediate') === 0;
    }

    static throwIfLarger(value, size, signed) {
        var val_size = signed ? Constant.sizeClass(value) : Constant.sizeClassUnsigned(value);
        if(val_size > size) throw TypeError(`Value ${value} too big for imm8.`);
    }

    variable: Variable = null;

    setVariable(variable: Variable) {
        this.variable = variable;
    }

    cast(ImmediateClass: typeof Immediate) {
        return new ImmediateClass(this.value);
    }

    toString() {
        if(this.variable) return this.variable.toString();
        else return super.toString();
    }
}

export class ImmediateUnsigned extends Immediate {
    constructor(value: number|number64 = 0) {
        super(value, false);
    }
}

export class Immediate8 extends Immediate {
    setValue(value: number|number64) {
        Immediate.throwIfLarger(value, SIZE.B, this.signed);
        super.setValue(value);
        this.extend(SIZE.B);
    }
}

export class ImmediateUnsigned8 extends Immediate8 {
    constructor(value: number|number64 = 0) {
        super(value, false);
    }
}

export class Immediate16 extends Immediate {
    setValue(value: number|number64) {
        Immediate.throwIfLarger(value, SIZE.W, this.signed);
        super.setValue(value);
        this.extend(SIZE.W);
    }
}

export class ImmediateUnsigned16 extends Immediate16 {
    constructor(value: number|number64 = 0) {
        super(value, false);
    }
}

export class Immediate32 extends Immediate {
    setValue(value: number|number64) {
        Immediate.throwIfLarger(value, SIZE.D, this.signed);
        super.setValue(value);
        this.extend(SIZE.D);
    }
}

export class ImmediateUnsigned32 extends Immediate32 {
    constructor(value: number|number64 = 0) {
        super(value, false);
    }
}

export class Immediate64 extends Immediate {
    setValue(value: number|number64) {
        Immediate.throwIfLarger(value, SIZE.Q, this.signed);
        super.setValue(value);
        this.extend(SIZE.Q);
    }
}

export class ImmediateUnsigned64 extends Immediate64 {
    constructor(value: number|number64 = 0) {
        super(value, false);
    }
}


// ## Registers
//
// `Register` represents one of `%rax`, `%rbx`, etc. registers.
export abstract class Register extends Operand {
    id: number = 0; // Number value of register.
    name: string = 'reg';

    constructor(id: number, size: SIZE) {
        super();
        this.id = id;
        this.size = size;
    }

    abstract ref(): Memory;

    reg(): Register {
        return this;
    }

    getName() {
        return this.name;
    }

    idSize() { // In bits
        if(this.id < 0b1000)    return 3;
        if(this.id < 0b10000)   return 4;
        if(this.id < 0b100000)  return 5;
        throw Error('Register ID too big.');
    }

    get3bitId() {
        return this.id & 0b111;
    }

    get4bitId() {
        return this.id & 0b1111;
    }

    toNumber() {
        return this.id;
    }

    toString() {
        return this.name;
    }
}


// ## Memory
//
// `Memory` is RAM addresses which `Register`s can *dereference*.
export class Memory extends Operand {

    base: Register = null;

    reg(): Register {
        return this.base;
    }

    getAddressSize(): SIZE {
        var reg = this.reg();
        if(reg) return reg.size;
        return SIZE.NONE;
    }

    getOperandSize(): SIZE {
        return this.size;
    }

    ref(base: Register): this {
        this.base = base;
        return this;
    }

    toString() {
        return `[${this.base.toString()}]`;
    }
}


// Operand which needs `evaluation`, it may be that it cannot evaluate on first two passes.
export class Variable extends Operand {
    result: Tnumber = null; // Result of evaluation.

    canEvaluate(owner: Expression): boolean {
        return true;
    }

    evaluate(owner: Expression): Tnumber {
        return 0;
    }

    // Evaluate approximately during 2nd pass.
    evaluatePreliminary(owner: Expression): Tnumber {
        return 0;
    }
}


export type Tvariable = Tnumber|Variable;
export function isTvariable(val) {
    if(val instanceof Variable) return true;
    else return isTnumber(val);
}


// Relative jump targets for jump instructions.
export class Relative extends Variable {
    static size = SIZE.ANY;

    static fromExpression(expr: Expression) {
        return new Relative(expr);
    }

    signed = true;

    target: Expression; // Target `Expression` which is referenced by this `Relative`.
    offset: number;

    constructor(target: Expression, offset: number = 0) {
        super();
        this.target = target;
        this.offset = offset;
    }

    canEvaluate(owner) {
        if(!owner || (owner.offset === -1)) return false;
        if(this.target.offset === -1) return false;
        return true;
    }

    evaluate(owner) {
        return this.result = this.rebaseOffset(owner) - owner.bytes();
        // return this.result = this.rebaseOffset(owner);
    }

    evaluatePreliminary(owner: Expression) {
        return this.offset + this.target.offsetMax - owner.offsetMax;
    }

    canHoldMaxOffset(owner: Expression) {
        var value = this.evaluatePreliminary(owner);
        var size = this.signed ? Constant.sizeClass(value) : Constant.sizeClassUnsigned(value);
        return size <= this.size;
    }

    clone(): Relative {
        return new Relative(this.target, this.offset);
    }

    cast(RelativeClass: typeof Relative): Relative {
    // cast(RelativeClass: typeof Relative) {
        this.size = RelativeClass.size;
        // return new RelativeClass(this.target, this.offset);
        return this;
    }

    rebaseOffset(new_target: Expression): number {
        // if(expr.code !== this.expr.code)
        //     throw Error('Rebase from different code blocks not implemented yet.');
        if(new_target.offset === -1)
            throw Error('Expression has no offset, cannot rebase.');
        return this.offset + this.target.offset - new_target.offset;
    }

    // Recalculate relative offset given a different Expression.
    // rebase(target: Expression): Relative {
    rebase(target: Expression) {
        // return new Relative(target, this.rebaseOffset(expr));
        this.offset = this.rebaseOffset(target);
        this.target = target;
    }

    toNumber() {
        return this.offset;
    }

    toString() {
        var result = '';
        if(this.result !== null) {
            result = ' = ' + this.result;
        }

        if(this.target instanceof require('./Label').default) {
            var lbl = this.target as Label;
            var off = this.offset ? '+' + (new Constant(this.offset)).toString() : '';
            return `<${lbl.getName()}${off}${result}>`;
        } else if(this.target.asm) {
            // var lbl = this.target.asm.getStartLabel();
            const lbl = this.target.asm.expressions[0] as any as Label;
            var expr = `+[${this.target.index}]`;
            var off = this.offset ? '+' + (new Constant(this.offset)).toString() : '';
            return `<${lbl.getName()}${expr}${off}${result}>`;
        } else {
            var expr = `+[${this.target.index}]`;
            var off = this.offset ? '+' + (new Constant(this.offset)).toString() : '';
            return `<${expr}${off}${result}>`;
        }
    }
}

export class Relative8 extends Relative {
    static size = SIZE.B;
}

export class Relative16 extends Relative {
    static size = SIZE.W;
}

export class Relative32 extends Relative {
    static size = SIZE.D;
}


export class Symbol extends Relative {
    private static cnt = 0;

    name: string;

    constructor(target: Expression, offset: number = 0, name?: string) {
        super(target, offset);
        this.name = name ? name : 'symbol_' + (Symbol.cnt++);
    }
}


export type TOperand = (Tnumber|Operand|Expression);

// Collection of operands an `Expression` might have.
export class Operands {

    static findSize(ops: TUiOperand[]): SIZE {
        for(var operand of ops) {
            if(operand instanceof Register) return (operand as Register).size;
        }
        return SIZE.NONE;
    }

    static uiOpsNormalize(ops: TUiOperand[]): TUiOperandNormalized[] {
        var i = require('./instruction');
        // Wrap `Expression` into `Relative`.
        for(var j = 0; j < ops.length; j++) {
            if(ops[j] instanceof Expression) {
                ops[j] = (ops[j] as Expression).rel();
            }
        }
        return ops as TUiOperandNormalized[];
    }

    list: TOperand[] = [];

    size: SIZE = SIZE.ANY; // Size of each operand.


    constructor(list: TOperand[] = [], size: SIZE = SIZE.ANY) {
        this.size = size;
        this.list = list;
    }

    clone(Clazz: typeof Operands = Operands) {
        var list = [];
        for(var op of this.list) list.push(op);
        return new Clazz(list, this.size);
    }

    // Wrap `Expression` into `Relative`.
    normalizeExpressionToRelative() {
        var i = require('./instruction');
        var ops = this.list;
        for(var j = 0; j < ops.length; j++) {
            if(ops[j] instanceof i.Expression) {
                ops[j] = (ops[j] as Expression).rel();
            }
        }
    }

    validateSize() {
        // Verify operand sizes.
        for(var op of this.list) {
            // We can determine operand size only by Register; Memory and Immediate and others don't tell us the right size.
            if(op instanceof Register) {
                if (this.size !== SIZE.ANY) {
                    // if (this.size !== op.size)
                    //     throw TypeError('Operand size mismatch.');
                } else this.setSize(op.size);
            }
        }
    }

    setSize(size: SIZE) {
        if(this.size === SIZE.ANY) this.size = size;
        else throw TypeError('Operand size mismatch.');
    }

    getAtIndexOfClass(index, Clazz) {
        var op = this.list[index];
        if(op instanceof Clazz) return op;
        else return null;
    }

    getFirstOfClass(Clazz, skip = 0) {
        for(var op of this.list) {
            if(op instanceof Clazz) {
                if(!skip) return op;
                else skip--;
            }
        }
        return null;
    }

    getRegisterOperand(skip = 0): Register {
        return this.getFirstOfClass(Register, skip) as Register;
    }

    getMemoryOperand(skip = 0): Memory {
        return this.getFirstOfClass(Memory, skip) as Memory;
    }

    getVariable(skip = 0): Variable {
        return this.getFirstOfClass(Variable, skip) as Variable;
    }

    getRelative(skip = 0): Relative {
        return this.getFirstOfClass(Relative, skip) as Relative;
    }

    getImmediate(skip = 0): Immediate {
        return this.getFirstOfClass(Immediate, skip) as Immediate;
    }

    getAddressSize(): SIZE {
        var mem = this.getMemoryOperand();
        if(mem) return mem.size;
        else return SIZE.NONE;
    }

    hasRegister(): boolean {
        return !!this.getRegisterOperand();
    }

    hasMemory(): boolean {
        return !!this.getMemoryOperand();
    }

    hasVariable(): boolean {
        return !!this.getMemoryOperand();
    }

    hasRelative(): boolean {
        return !!this.getRelative();
    }

    hasRegisterOrMemory(): boolean {
        return this.hasRegister() || this.hasMemory();
    }

    canEvaluate(owner: Expression): boolean {
        for(var op of this.list) {
            if(op instanceof Variable)
                if(!(op as Variable).canEvaluate(owner)) return false;
        }
        return true;
    }

    evaluate(owner: Expression) {
        for(var op of this.list)
            if(op instanceof Variable) (op as Variable).evaluate(owner);
    }

    // EVEX may encode up to 4 operands, 32 registers, so register can be up to 5-bits wide,
    // we need to check for that because in that case we cannot use VEX.
    has5bitRegister() {
        for(var j = 0; j < 4; j++) {
            var op = this.list[j];
            if(!op) break;
            if(op instanceof Register) {
                if((op as Register).idSize() > 4) return true;
            }
        }
        return false;
    }

    toString() {
        return this.list.map((op) => { return op.toString(); }).join(', ');
    }
}
