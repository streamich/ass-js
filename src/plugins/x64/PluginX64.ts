import Plugin from "../Plugin";
import {InstructionX64} from './instruction';


class PluginX64 extends Plugin {
    onAsm (asm) {
        asm.hooks.instruction.tap('PluginX64', () => new InstructionX64());
    }
}

export default PluginX64;
