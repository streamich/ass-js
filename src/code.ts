import {SIZE, number64, Tnumber, Operands} from './operand';
import {Expression, Label, Data, DataUninitialized, DataVariable} from './instruction';
import * as i from './instruction';
import * as o from './operand';
import * as d from './def';
import {UInt64, extend} from './util';





export class Code {
    expr: Expression[] = [];

    ClassInstruction = i.Instruction;
    ClassInstructionSet = i.InstructionSet;
    ClassOperands = o.Operands;

    littleEndian = true; // Which way to encode constants by default.

    table: d.DefTable;



    // Expressions are compiled in 3 passes:
    //
    //  - *1st pass* -- maximum offset `maxOffset` for each expression is computed, some expression might not know
    //  their size jet, not all expressions are known, future references. First pass is when user performs insertion of commands.
    //  - *2nd pass* -- all expressions known now, each expression should pick its right size, exact `offset` is computed for each expression.
    //  - *3rd pass* -- now we know exact `offset` of each expression, so in this pass we fill in the addresses.
    compile(): number[] {
        // 1st pass is performed as instructions are `insert`ed, `.offsetMax` is calculated, and possibly `.offset`.

        // Instructions without size can now determine their size based on `.offsetMax` and
        // calculate their real `.offset`.
        this.do2ndPass();

        // Offsets are now know, here we evaluate references.
        return this.do3rdPass();
    }

    do2ndPass() {
        // We probably cannot skip this 2nd pass, as instructions might change their sizes after inserted,
        // for example, when `.lock()` prefix is added.
        // var last = this.expr[this.expr.length - 1];
        // var all_offsets_known = last.offset >= 0;
        //
        // Edge case when only the last Expression has variable size.
        // var all_sizes_known = last.bytes() >= 0;
        //
        // if(all_offsets_known && all_sizes_known) return; // Skip 2nd pass.

        var prev = this.expr[0];
        prev.offset = 0;
        for(var j = 1; j < this.expr.length; j++) {
            var ins = this.expr[j];

            if(ins instanceof i.ExpressionVolatile) {
                var fixed = (ins as i.ExpressionVolatile).getFixedSizeExpression();
                this.replace(fixed, ins.index);
                ins = fixed;
                // (ins as i.ExpressionVolatile).determineSize();
            }

            // var bytes = prev.bytes();
            // if(bytes === i.SIZE_UNKNOWN)
            //     throw Error(`Instruction [${j}] does not have size.`);
            // ins.offset = prev.offset + bytes;

            // Need to call method, as `InstructionSet` contains multiple `Instruction`s,
            // that all need offset updated of picked instruction.
            ins.calcOffset();

            prev = ins;
        }
    }

    do3rdPass() {
        var code: number[] = [];
        for(var ins of this.expr) {
            if(ins instanceof i.ExpressionVariable)
                (ins as i.ExpressionVariable).evaluate();
            code = ins.write(code); // 3rd pass
        }
        return code;
    }
}
