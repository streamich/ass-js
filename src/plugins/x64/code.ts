import {SIZE, TUiOperand} from '../../operand';
import * as code from '../x86/code';
import * as d from '../x86/def';
import {InstructionX64} from './instruction';
import * as t from './table';
import {extend} from '../../util';


export class Code extends code.Code {

    static table = (new d.DefTable(t.table as any, t.defaults as any)) as any;

    private static _methodsAdded = false;

    static create(name: string = 'start') {
        var newcode = new Code(name);
        newcode.addMethods();
        return newcode;
    }

    table = Code.table;

    ClassInstruction = InstructionX64;

}
