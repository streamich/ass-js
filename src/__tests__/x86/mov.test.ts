import {rax, rbx, rcx, rdx, rbp, rsp, rsi, rdi, r8, r9, r10, r11, r12, r13, r15,
    ebx, eax, esp, ebp, ecx,
    r10d, ax, bx, cx, dx, al, bl, cl, dl, ah, ch, dil, bpl, spl, r8b, r10b,
    rip} from '../../plugins/x86/operand';
import * as o from '../../plugins/x86/operand';
import {Code} from '../../plugins/x64/code';


describe('mov', () => {
    let _: Code;

    beforeEach(() => {
        _ = new Code;
    });

    afterEach(() => {
        expect(_.compile()).toMatchSnapshot();
    });

    describe('capture snapshots', () => {
        it('movq rax, rax', () => {
            _._('mov', [rax, rax], 64);
        });
        it('movq rbx, rax', () => {
            _._('mov', [rbx, rax], 64);
        });
        it('movq [rax], rax', () => {
            _._('mov', [rax.ref(), rax], 64);
        });
        it('movq [rcx], rbx', () => {
            _._('mov', [rcx.ref(), rbx], 64);
        });
        it('movq rdx, [rbx]', () => {
            _._('mov', [rdx, rbx.ref()], 64);
        });
        it('movq r9, r8', () => {
            _._('mov', [r9, r8], 64);
        });
        it('movq [r11], r10', () => {
            _._('mov', [r11.ref(), r10], 64);
        });
        it('movq r13, [r12]', () => {
            _._('mov', [r13, r12.ref()], 64);
        });
        it('movq rcx, [rbx + 0x11]', () => {
            _._('mov', [rcx, rbx.disp(0x11)], 64);
        });
        it('movq rsi, [rcx + rdx]', () => {
            _._('mov', [rsi, rcx.ref().ind(rdx, 1)], 64);
        });
        it('movq rcx, [rax + rbx * 4 + 0x1234]', () => {
            _._('mov', [rcx, rax.ref().ind(rbx, 4).disp(0x1234)], 64);
        });
        it('movq rbp, [0x1]', () => {
            _._('mov', [rbp, _.mem(0x1)], 64);
        });
        it('movq rsp, [0x1]', () => {
            _._('mov', [rsp, _.mem(0x1)], 64);
        });
        it('movq [rsp], rbx', () => {
            _._('mov', [rsp.ref(), rbx], 64);
        });
        it('movq rsp, rbx', () => {
            _._('mov', [rsp, rbx], 64);
        });
        it('movq [rbp], rbx', () => {
            _._('mov', [rbp.ref(), rbx], 64);
        });
        it('movq [rsp + rbp * 2], rbx', () => {
            _._('mov', [rsp.ref().ind(rbp, 2), rbx], 64);
        });
        it('movq rbx, [rbp + rax * 8]', () => {
            _._('mov', [rbx, rbp.ref().ind(rax, 8)], 64);
        });
        it('movq rbp, [rdx * 2]', () => {
            _._('mov', [rbp, rdx.ind(2)], 64);
        });
        it('movq [rbp * 8 + -0x123], rsp', () => {
            _._('mov', [rbp.ind(8).disp(-0x123), rsp], 64);
        });
        it('movq rax, 0x1', () => {
            _._('mov', [rax, 0x1], 64);
        });
        // 48 c7 c5 cd cc ff ff 	movq $-0x3333, %rbp
        it('movq rax, 0x1', () => {
            _._('mov', [rbp, -0x3333], 64);
        });
        // 4c 89 eb             	mov    %r13,%rbx
        it('movq rbx, r13', () => {
            _._('mov', [rbx, r13], 64);
        });
    });
});
