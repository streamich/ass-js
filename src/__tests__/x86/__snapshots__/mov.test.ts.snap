// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mov capture snapshots movq [r11], r10 1`] = `
Array [
  77,
  137,
  19,
]
`;

exports[`mov capture snapshots movq [rax], rax 1`] = `
Array [
  72,
  137,
  0,
]
`;

exports[`mov capture snapshots movq [rax], rax 2`] = `
Array [
  72,
  137,
  0,
]
`;

exports[`mov capture snapshots movq [rbp * 8 + -0x123], rsp 1`] = `
Array [
  72,
  137,
  36,
  237,
  221,
  254,
  255,
  255,
]
`;

exports[`mov capture snapshots movq [rbp], rbx 1`] = `
Array [
  72,
  137,
  93,
  0,
]
`;

exports[`mov capture snapshots movq [rcx], rbx 1`] = `
Array [
  72,
  137,
  25,
]
`;

exports[`mov capture snapshots movq [rsp + rbp * 2], rbx 1`] = `
Array [
  72,
  137,
  28,
  108,
]
`;

exports[`mov capture snapshots movq [rsp], rbx 1`] = `
Array [
  72,
  137,
  28,
  36,
]
`;

exports[`mov capture snapshots movq r9, r8 1`] = `
Array [
  77,
  137,
  193,
]
`;

exports[`mov capture snapshots movq r13, [r12] 1`] = `
Array [
  77,
  139,
  44,
  36,
]
`;

exports[`mov capture snapshots movq rax, 0x1 1`] = `
Array [
  72,
  199,
  192,
  1,
  0,
  0,
  0,
]
`;

exports[`mov capture snapshots movq rax, 0x1 2`] = `
Array [
  72,
  199,
  197,
  205,
  204,
  255,
  255,
]
`;

exports[`mov capture snapshots movq rax, rax 1`] = `
Array [
  72,
  137,
  192,
]
`;

exports[`mov capture snapshots movq rbp, [0x1] 1`] = `
Array [
  72,
  139,
  44,
  37,
  1,
  0,
  0,
  0,
]
`;

exports[`mov capture snapshots movq rbp, [rdx * 2] 1`] = `
Array [
  72,
  139,
  44,
  85,
  0,
  0,
  0,
  0,
]
`;

exports[`mov capture snapshots movq rbx, [rbp + rax * 8] 1`] = `
Array [
  72,
  139,
  92,
  197,
  0,
]
`;

exports[`mov capture snapshots movq rbx, r13 1`] = `
Array [
  76,
  137,
  235,
]
`;

exports[`mov capture snapshots movq rbx, rax 1`] = `
Array [
  72,
  137,
  195,
]
`;

exports[`mov capture snapshots movq rcx, [rax + rbx * 4 + 0x1234] 1`] = `
Array [
  72,
  139,
  140,
  152,
  52,
  18,
  0,
  0,
]
`;

exports[`mov capture snapshots movq rcx, [rbx + 0x11] 1`] = `
Array [
  72,
  139,
  75,
  17,
]
`;

exports[`mov capture snapshots movq rdx, [rbx] 1`] = `
Array [
  72,
  139,
  19,
]
`;

exports[`mov capture snapshots movq rsi, [rcx + rdx] 1`] = `
Array [
  72,
  139,
  52,
  17,
]
`;

exports[`mov capture snapshots movq rsp, [0x1] 1`] = `
Array [
  72,
  139,
  36,
  37,
  1,
  0,
  0,
  0,
]
`;

exports[`mov capture snapshots movq rsp, rbx 1`] = `
Array [
  72,
  137,
  220,
]
`;
